File: Backend/fastApiProject/service/disaster_service.py
from geoalchemy2 import WKTElement
from sqlalchemy import func
from sqlalchemy.orm import Session
from messaging.publisher import EventPublisher
from model import Location, Disaster
from repository.interfaces.disaster_repo_interface import IDisasterRepository
from schema.request.disaster_create_request import DisasterCreateRequest
from schema.request.location_create_request import LocationCreateRequest
from service.location_service import LocationService


class DisasterService:
    def __init__(self,
                 disaster_repo: IDisasterRepository,
                 location_service: LocationService,
                 publisher: EventPublisher,
                 db: Session
    ):
        self.disaster_repo = disaster_repo
        self.location_service = location_service
        self.publisher = publisher
        self.db = db

    def create_disaster_manually(self, disaster: DisasterCreateRequest, user_id: int):
        location_dto = LocationCreateRequest(
            latitude=disaster.latitude,
            longitude=disaster.longitude,
            city=disaster.city,
            country=disaster.country
        )
        loc = self.location_service.create_location(location_dto, self.db)
        disaster = self.disaster_repo.create_disaster_manually(disaster, user_id, loc.id, self.db)

        payload = {
            'id': disaster.id,
            'title': disaster.title,
            'severity': disaster.severity,
            'latitude': loc.latitude,
            'longitude': loc.longitude
        }

        self.publisher.publish_disaster_created(payload)

        return disaster

    def get_disaster(self, disaster_id: int):
        return self.disaster_repo.get_disaster(disaster_id, self.db)

    def persist_disasters_third_party(self, disasters):
        for disaster in disasters:
            location = Location(
                latitude=disaster.latitude,
                longitude=disaster.longitude,
                city=disaster.city,
                country=disaster.country,
                coordinates=WKTElement(f"POINT({disaster.longitude} {disaster.latitude})", srid=4326)
            )

            if self.disaster_repo.external_id_exists(disaster.external_id, self.db):
                continue
            try:

                location = self.location_service.create_location_no_commit(location, self.db)

                new_disaster = Disaster(
                    title=disaster.title,
                    description=disaster.description,
                    type=disaster.type,
                    severity=disaster.severity,
                    source=disaster.source,
                    status=disaster.status,
                    radius=disaster.radius,
                    location_id=location.id,
                    external_id=disaster.external_id,
                    start_time=disaster.start_time,
                    end_time=disaster.end_time
                )
                new_disaster = self.disaster_repo.create_disaster_no_commit(new_disaster, self.db)

                self.db.commit()

                payload = {
                    'id': new_disaster.id,
                    'title': new_disaster.title,
                    'severity': new_disaster.severity,
                    'latitude': location.latitude,
                    'longitude': location.longitude
                }

                self.publisher.publish_disaster_created(payload)
                # print("Created " + disaster.title)
            except Exception as exp:
                # print(exp)
                self.db.rollback()

    def get_disaster_nearby(self, lat, lng, rad, sev, typ):
        return self.disaster_repo.get_disaster_nearby(lat, lng, rad, sev, typ, self.db)

File: disaster-live-alerts/src/main/java/com/example/disaster_live_alerts/service/impl/DisasterService.java
package com.example.disaster_live_alerts.service.impl;

import com.example.disaster_live_alerts.dto.*;
import com.example.disaster_live_alerts.enums.DisasterSeverity;
import com.example.disaster_live_alerts.enums.DisasterSource;
import com.example.disaster_live_alerts.enums.DisasterType;
import com.example.disaster_live_alerts.model.Disaster;
import com.example.disaster_live_alerts.model.Location;
import com.example.disaster_live_alerts.repo.DisasterRepository;
import com.example.disaster_live_alerts.security.CustomUserDetails;
import com.example.disaster_live_alerts.service.IDisasterService;
import com.example.disaster_live_alerts.service.ILocationService;
import com.example.disaster_live_alerts.websocket.LiveAlertsWebSocketHandler;
import jakarta.transaction.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
public class DisasterService implements IDisasterService {
    private final DisasterRepository disasterRepository;
    private final ILocationService locationService;
    private final LiveAlertsWebSocketHandler liveAlertsWebSocketHandler;

    public DisasterService(DisasterRepository disasterRepository, ILocationService locationService, LiveAlertsWebSocketHandler liveAlertsWebSocketHandler) {
        this.disasterRepository = disasterRepository;
        this.locationService = locationService;
        this.liveAlertsWebSocketHandler = liveAlertsWebSocketHandler;
    }

    @Override
    @Transactional
    public DisasterCreateResponseDto createDisasterManually(DisasterCreateRequestDto disasterCreateRequestDto) {

        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        CustomUserDetails userDetails = (CustomUserDetails) auth.getPrincipal();
        if (userDetails == null) {
            throw new RuntimeException("User Principal is null");
        }

        LocationDto locationDto =
                LocationDto.builder()
                        .longitude(disasterCreateRequestDto.getLongitude())
                        .latitude(disasterCreateRequestDto.getLatitude())
                        .country(disasterCreateRequestDto.getCountry())
                        .city(disasterCreateRequestDto.getCity()).build();

        Location loc = locationService.createLocation(locationDto);

        Disaster disaster =
                Disaster.builder()
                        .title(disasterCreateRequestDto.getTitle())
                        .description(disasterCreateRequestDto.getDescription())
                        .type(disasterCreateRequestDto.getType())
                        .location(loc)
                        .radius(disasterCreateRequestDto.getRadius())
                        .createdBy(userDetails.getCurrentUser())
                        .severity(disasterCreateRequestDto.getSeverity())
                        .status(disasterCreateRequestDto.getStatus())
                        .startTime(disasterCreateRequestDto.getStartTime())
                        .source(DisasterSource.MANUAL)
                        .endTime(disasterCreateRequestDto.getEndTime()).build();

        disasterRepository.save(disaster);

        NormalizedDisasterDto dto =
                new NormalizedDisasterDto(
                        disaster.getTitle(),
                        disaster.getDescription(),
                        disaster.getType(),
                        disaster.getSeverity(),
                        disaster.getSource(),
                        disaster.getStatus(),
                        disaster.getRadius(),
                        null,
                        disaster.getStartTime(),
                        locationDto.getLatitude(),
                        locationDto.getLongitude(),
                        locationDto.getCity(),
                        locationDto.getCountry()
                );

        liveAlertsWebSocketHandler.broadcastAlert(dto, 10000);

        return DisasterCreateResponseDto.builder()
                .id(disaster.getId())
                .title(disaster.getTitle())
                .description(disaster.getDescription())
                .type(disaster.getType())
                .severity(disaster.getSeverity())
                .status(disaster.getStatus())
                .radius(disaster.getRadius())
                .startTime(disaster.getStartTime())
                .endTime(disaster.getEndTime()).build();
    }

    @Override
    public Page<DisasterNearbyResponseDto> getNearbyDisasters(Double lat,
                                                              Double lng,
                                                              Double rad,
                                                              DisasterSeverity sev,
                                                              DisasterType typ,
                                                              Pageable pageable) {
        return disasterRepository.getNearbyDisasters(lat,lng, rad, sev, typ, pageable);
    }

    @Override
    public DisasterFetchResponseDto getDisaster(Long disasterId) {
        return disasterRepository.getDisasterById(disasterId)
                .orElseThrow(() -> new RuntimeException("Disaster not found"));
    }

    @Override
    @Transactional
    public void createDisasterFromThirdParty(NormalizedDisasterDto disasterDto) {
        if (disasterRepository.existsByExternalId(disasterDto.getExternalId())) {
            return;
        }
        LocationDto loc =
                LocationDto.builder()
                        .longitude(disasterDto.getLongitude())
                        .latitude(disasterDto.getLatitude())
                        .city(disasterDto.getCity())
                        .country(disasterDto.getCountry()).build();

        Location location = locationService.createLocation(loc);

        Disaster disaster =
                Disaster.builder()
                        .title(disasterDto.getTitle())
                        .description(disasterDto.getDescription())
                        .type(disasterDto.getType())
                        .severity(disasterDto.getSeverity())
                        .source(disasterDto.getSource())
                        .status(disasterDto.getStatus())
                        .radius(disasterDto.getRadius())
                        .externalId(disasterDto.getExternalId())
                        .startTime(disasterDto.getStartTime())
                        .location(location).build();

        disasterRepository.save(disaster);
    }
}
